### ISSUE TITLE
k3os.io/upgrade from dev release will downgrade to stable release

### ISSUE THREAD

**Issue by sob** [rancher/k3os#642]
**Version (k3OS / kernel)**
k3os version v0.19.4-dev.5
5.4.79-v8+ #1373 SMP PREEMPT Mon Nov 23 13:32:41 GMT 2020

**Architecture**
aarch64

**Describe the bug**
Creating a new cluster using v0.19.4-dev.5 with node labels k3os.io/upgrade will downgrade to the latest stable release (v0.11.1 with v1.18.9+k3s1)

**To Reproduce**

1. Provision a node with v0.19.4-dev5 and the following in k3os config.yaml
```
k3os:
  labels:
    k3os.io/upgrade: enabled
```
2. After nodes boot up, verify that system-upgrade controller is running
3. Verify that the k3os --version reports v0.19.4-dev.5 
4. system-upgrade controller will actually downgrade them to the latest stable release v0.11.1 (v1.18.9+k3s1).

**Expected behavior**
system-upgrade controller to detect a higher version number and not upgrade, or better yet allow a label option to upgrade to pre-releases versions.

**Actual behavior**
system-upgrade controller downgrades nodes to the latest stable release (v0.11.1 v1.18.9+k3s1).

**Additional context**
Wasn't sure if it was my configuration or if there was a known workaround but I couldn't seem to find any references to additional configuration options in the labels.

**Comment by nsmith5** [comment]
Agreed. I found this behaviour a little annoying. If I can add one more suggestion: It would be nice to give a specific upgrade version instead of relying on what latest might be at a given time to make upgrades reproducible.

**Comment by dweomer** [comment]
> Agreed. I found this behaviour a little annoying. If I can add one more suggestion: It would be nice to give a specific upgrade version instead of relying on what latest might be at a given time to make upgrades reproducible.

This is within your control and why the supplied plan for the upgrade controller is disabled by default. Currently k3OS lacks a channel server similar to what we have for k3s. This means that channel management (and configuration in the plan) is an exercise left to the human operator. That said, you can easily pin the version to a dev release by specifying either:
```yaml
spec:
  channel: "https://github.com/rancher/k3os/releases/tag/v0.19.4-dev.5"
  # or
  version: v0.19.4-dev.5
```

**Comment by nsmith5** [comment]
Ah ha! I should have known that was possible.

A silly mistake I'll leave here for future readers: _Best to not edit the existing plan but make a new one instead_. 

As you can imagine, if you update the channel or version on the existing `k3os-latest` plan and upgrade your control plane node you'll find the `k3os-latest` plan is reset to the original. This will result in your node updating and then reverting back right after. Better to create a new plan and use a different label selection (e.g `example.com/upgrade=enabled`). Three are probably other solutions to this as well..

Thanks again for the insight @dweomer !

**Comment by Alxandr** [comment]
Related question; if I unset the upgrade label on a running cluster that has the update label in config, will it be re-added after the upgrade? (so downgrade will happen, and then upgrade and then downgrade forever?)

**Comment by dweomer** [comment]
> Related question; if I unset the upgrade label on a running cluster that has the update label in config, will it be re-added after the upgrade? (so downgrade will happen, and then upgrade and then downgrade forever?)

@Alxandr my apologies for missing this until now, 3 weeks later: if you remove the `plan.upgrade.cattle.io/${planName}` from a node (or set the value to `disabled`) then the upgrade controller will no longer consider that node for upgrades and therefore no upgrades will be applied.

**Comment by Alxandr** [comment]
My worry is that the label will be re-added on the next boot (given that it is contained in the configuration file).

**Comment by dweomer** [comment]
> My worry is that the label will be re-added on the next boot (given that it is contained in the configuration file).

Yes, that will happen. Will it upgrade-then-downgrade forever because of this? With a properly configured plan, as mentioned at https://github.com/rancher/k3os/issues/642#issuecomment-754918742, it shouldn't.
